FirebaseDatabase/Sources/Utilities/FNextPushId.m

```diff
@@ -93,19 +100,91 @@ + (NSString *)successor:(NSString *_Nonnull)key {
         return [FUtilities maxName];
     }
 
-    NSString *source =
-        [NSString stringWithFormat:@"%C", [next characterAtIndex:i]];
-    NSInteger sourceIndex = [PUSH_CHARS rangeOfString:source].location;
-    NSString *sourcePlusOne = [NSString
-        stringWithFormat:@"%C", [PUSH_CHARS characterAtIndex:sourceIndex + 1]];
+    unichar character = [next characterAtIndex:i];
+    unichar plusOne = character + 1;
+    BOOL removePreviousCharacter = NO;
+    BOOL replaceWithLowestSurrogatePair = NO;
+    switch (plusOne) {
+    case 0x23: // '#'
+    case 0x24: // '$'
+        plusOne = 0x25;
+        break;
+
+    case 0x2E: // '.'
+    case 0x2F: // '/'
+        plusOne = 0x30;
+        break;
+
+    case 0x5B: // '['
+        plusOne = 0x5C;
+        break;
+
+    case 0x5D: // ']'
+        plusOne = 0x5E;
+        break;
 
-    [next replaceCharactersInRange:NSMakeRange(i, i + 1)
+    case 0x7F: // control character: del
+        plusOne = 0x80;
+        break;
+
+    case HIGH_SURROGATE_PAIR_START: // 0xD800
+        // We added one to 0xD7FF and entered surrogate pair zone
```

Comment:
To me, this logic feels a bit hard to parse since you are operating on the already mutated value. Do you think it is possible to use the base value and then increment it based on these special cases?

Reply:
Right, so you mean make the entire switch statement operate on the `character` variable rather than on `plusOne`? That can of course be done, but switching on `plusOne` makes the values in the various cases exactly match the 'invalid' ones - or the ones where the `plusOne` is not the next value, but must be replaced by something else.

But I agree that the logic is both hard to read and it was also not easy to write.

Reply:
I will play with this today and let you know if I can come up with something clean. Thanks for your patience.

Reply:
Okay, so after two hours of playing with this I have now managed to understand why you wrote it this way. It is much easier to deal with the "forbidden" codepoints once translated. My understanding of Unicode is a bit murky, so I am not 100% convinced that the logic below works in all cases - but I think we can merge it, especially since it will be short lived. 

It would be good to add a test for the successor method similar to `testPredessorUnicode()`. In the meantime, I will try to see if I can improve my understanding.

Reply:
I'm no Unicode expert at all, but luckily all we need to know is about utf-16 encodings of Unicode code points, and that seems very well defined. My source of information on that actually just comes from the Wikipedia entry for utf-16.